File: crates/nexushub-core/tests/contract_registry.rs
use nexushub_core::platform::{PlatformKind, PlatformPaths};
use nexushub_core::services::{
    commands,
    contract_dtos::{CONTRACT_DTO_NAMES, CONTRACT_DTO_OWNER_NAMES},
    system::{Capability, HostSurface},
};
use serde_json::Value;
use std::collections::BTreeSet;

fn contract() -> Value {
    let path = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../contracts/nexushub-contract.json"
    );
    let text = std::fs::read_to_string(path)
        .unwrap_or_else(|err| panic!("contract registry must exist at {path}: {err}"));
    serde_json::from_str(&text).expect("contract registry must be valid JSON")
}

fn contract_schema() -> Value {
    let path = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../contracts/nexushub-contract.schema.json"
    );
    let text = std::fs::read_to_string(path)
        .unwrap_or_else(|err| panic!("contract schema must exist at {path}: {err}"));
    serde_json::from_str(&text).expect("contract schema must be valid JSON")
}

fn string_set(value: &Value, key: &str) -> BTreeSet<String> {
    value
        .get(key)
        .and_then(Value::as_array)
        .unwrap_or_else(|| panic!("contract key {key} must be an array"))
        .iter()
        .map(|item| {
            item.as_str()
                .unwrap_or_else(|| panic!("contract key {key} entries must be strings"))
                .to_string()
        })
        .collect()
}

fn nested_string_set(value: &Value, object_key: &str, key: &str) -> BTreeSet<String> {
    value
        .get(object_key)
        .unwrap_or_else(|| panic!("contract key {object_key} must exist"))
        .get(key)
        .and_then(Value::as_array)
        .unwrap_or_else(|| panic!("contract key {object_key}.{key} must be an array"))
        .iter()
        .map(|item| {
            item.as_str()
                .unwrap_or_else(|| {
                    panic!("contract key {object_key}.{key} entries must be strings")
                })
                .to_string()
        })
        .collect()
}

fn action_set(contract: &Value, predicate: impl Fn(&Value) -> bool) -> BTreeSet<String> {
    contract
        .get("actions")
        .and_then(Value::as_array)
        .expect("contract actions must be an array")
        .iter()
        .filter(|action| predicate(action))
        .map(|action| {
            action
                .get("id")
                .and_then(Value::as_str)
                .expect("contract action id must be a string")
                .to_string()
        })
        .collect()
}

#[test]
fn contract_schema_locks_registry_top_level_shape() {
    let schema = contract_schema();
    let required = string_set(&schema, "required");
    let expected = [
        "schemaVersion",
        "hostSurfaces",
        "capabilities",
        "capabilitiesByHostSurface",
        "visual",
        "actions",
        "dtoCatalog",
    ]
    .into_iter()
    .map(str::to_string)
    .collect::<BTreeSet<_>>();
    assert_eq!(required, expected);

    let properties = schema
        .get("properties")
        .and_then(Value::as_object)
        .expect("contract schema properties must be an object")
        .keys()
        .cloned()
        .collect::<BTreeSet<_>>();
    assert_eq!(properties, expected);

    let action_required = schema
        .pointer("/$defs/action/required")
        .and_then(Value::as_array)
        .expect("contract schema must define required action fields")
        .iter()
        .map(|item| {
            item.as_str()
                .expect("required action field must be a string")
                .to_string()
        })
        .collect::<BTreeSet<_>>();
    assert_eq!(
        action_required,
        ["id", "kind", "scope", "coreUseCase"]
            .into_iter()
            .map(str::to_string)
            .collect::<BTreeSet<_>>()
    );
}

#[test]
fn contract_registry_matches_core_host_surfaces_and_capabilities() {
    let contract = contract();
    let expected_surfaces = HostSurface::ALL
        .iter()
        .map(|surface| surface.as_str().to_string())
        .collect::<BTreeSet<_>>();
    let expected_capabilities = Capability::all()
        .iter()
        .map(|capability| capability.as_str().to_string())
        .collect::<BTreeSet<_>>();

    assert_eq!(string_set(&contract, "hostSurfaces"), expected_surfaces);
    assert_eq!(string_set(&contract, "capabilities"), expected_capabilities);
}

#[test]
fn contract_registry_capability_matrix_matches_core_policy() {
    let contract = contract();
    for (surface, platform) in [
        (
            HostSurface::LinuxServerWebui,
            PlatformPaths::for_kind(PlatformKind::Linux),
        ),
        (
            HostSurface::DesktopEmbeddedTauri,
            PlatformPaths::for_kind(PlatformKind::Macos),
        ),
        (
            HostSurface::DesktopLanWebui,
            PlatformPaths::for_kind(PlatformKind::Macos),
        ),
    ] {
        let expected = Capability::all()
            .iter()
            .filter(|capability| capability.is_supported_on_surface(&platform, surface))
            .map(|capability| capability.as_str().to_string())
            .collect::<BTreeSet<_>>();
        assert_eq!(
            nested_string_set(&contract, "capabilitiesByHostSurface", surface.as_str()),
            expected,
            "contract capability matrix drifted for {}",
            surface.as_str()
        );
    }
}

#[test]
fn contract_registry_covers_all_linux_rpc_and_transport_commands() {
    let contract = contract();
    let rpc_actions = action_set(&contract, |action| {
        action
            .get("linuxRpc")
            .and_then(Value::as_str)
            .map(|value| !value.is_empty())
            .unwrap_or(false)
    });
    let transport_actions = action_set(&contract, |action| {
        action.get("kind").and_then(Value::as_str) == Some("transport")
    });

    let expected_rpc = commands::ALLOWED_RPC_COMMANDS
        .iter()
        .map(|command| command.to_string())
        .collect::<BTreeSet<_>>();
    let expected_transport = commands::ALLOWED_TRANSPORT_COMMANDS
        .iter()
        .map(|command| command.to_string())
        .collect::<BTreeSet<_>>();

    assert_eq!(rpc_actions, expected_rpc);
    assert_eq!(transport_actions, expected_transport);
}

#[test]
fn shared_actions_declare_core_webui_linux_rpc_and_tauri_mappings() {
    let contract = contract();
    let actions = contract
        .get("actions")
        .and_then(Value::as_array)
        .expect("contract actions must be an array");

    for action in actions {
        let id = action
            .get("id")
            .and_then(Value::as_str)
            .expect("contract action id must be a string");
        let kind = action
            .get("kind")
            .and_then(Value::as_str)
            .expect("contract action kind must be a string");
        let scope = action
            .get("scope")
            .and_then(Value::as_str)
            .expect("contract action scope must be a string");
        if let Some(linux_rpc) = action.get("linuxRpc").and_then(Value::as_str) {
            assert_eq!(
                linux_rpc, id,
                "contract action {id} must not introduce a second Linux RPC name"
            );
        }
        match scope {
            "shared" => {
                for key in [
                    "coreUseCase",
                    "linuxRpc",
                    "tauriCommand",
                    "webuiWrapper",
                    "dtoOwner",
                    "requestDto",
                    "responseDto",
                ] {
                    assert!(
                        action
                            .get(key)
                            .and_then(Value::as_str)
                            .map(|value| !value.trim().is_empty())
                            .unwrap_or(false),
                        "shared action {id} must declare {key}"
                    );
                }
            }
            "host_only" => {
                assert!(
                    action
                        .get("hostOnlyReason")
                        .and_then(Value::as_str)
                        .map(|value| !value.trim().is_empty())
                        .unwrap_or(false),
                    "host-only action {id} must declare hostOnlyReason"
                );
            }
            "transport" => {
                assert_eq!(kind, "transport", "transport scope must use transport kind");
                for key in ["webuiWrapper", "dtoOwner", "requestDto", "responseDto"] {
                    assert!(
                        action
                            .get(key)
                            .and_then(Value::as_str)
                            .map(|value| !value.trim().is_empty())
                            .unwrap_or(false),
                        "transport action {id} must declare {key}"
                    );
                }
            }
            other => panic!("unsupported action scope {other} for {id}"),
        }
    }
}

#[test]
fn contract_registry_dto_catalog_matches_core_markers() {
    let contract = contract();
    let actions = contract
        .get("actions")
        .and_then(Value::as_array)
        .expect("contract actions must be an array");
    let dto_catalog = contract
        .get("dtoCatalog")
        .and_then(Value::as_object)
        .expect("contract dtoCatalog must be an object");
    let core_dtos = CONTRACT_DTO_NAMES
        .iter()
        .map(|name| name.to_string())
        .collect::<BTreeSet<_>>();
    let contract_dtos = dto_catalog
        .values()
        .map(|entry| {
            entry
                .get("core")
                .and_then(Value::as_str)
                .expect("contract dtoCatalog entry must declare core")
                .to_string()
        })
        .collect::<BTreeSet<_>>();
    assert_eq!(
        core_dtos, contract_dtos,
        "core DTO marker list must match contracts/nexushub-contract.json dtoCatalog"
    );

    let core_owners = CONTRACT_DTO_OWNER_NAMES
        .iter()
        .map(|name| name.to_string())
        .collect::<BTreeSet<_>>();
    for action in actions {
        let id = action
            .get("id")
            .and_then(Value::as_str)
            .expect("contract action id must be a string");
        let scope = action
            .get("scope")
            .and_then(Value::as_str)
            .expect("contract action scope must be a string");
        if scope != "shared" && scope != "transport" {
            continue;
        }
        let owner = action
            .get("dtoOwner")
            .and_then(Value::as_str)
            .unwrap_or_else(|| panic!("contract action {id} must declare dtoOwner"));
        assert!(
            core_owners.contains(owner),
            "contract action {id} uses unknown DTO owner {owner}"
        );
        for key in ["requestDto", "responseDto"] {
            let dto_name = action
                .get(key)
                .and_then(Value::as_str)
                .unwrap_or_else(|| panic!("contract action {id} must declare {key}"));
            assert!(
                dto_catalog.contains_key(dto_name),
                "contract action {id} references unknown DTO {dto_name}"
            );
        }
    }
}

File: webui/src/lib/domain/contractRegistry.ts
import registry from "../../../../contracts/nexushub-contract.json";
import type { HostSurface } from "../../types";

export type ContractActionScope = "shared" | "host_only" | "transport";
export type ContractActionKind = "rpc" | "transport";

export type ContractAction = {
  id: string;
  kind: ContractActionKind;
  scope: ContractActionScope;
  coreUseCase?: string;
  linuxRpc?: string;
  tauriCommand?: string;
  webuiWrapper?: string;
  dtoOwner?: string;
  requestDto?: string;
  responseDto?: string;
  hostOnlyReason?: string;
};

export type ContractDtoCatalogEntry = {
  core: string;
  webui: string;
};

export type ContractVisual = {
  navigation: string[];
  corePanelTitles: string[];
  actionLabels: Record<string, string>;
  disabledStates: Record<string, string>;
  linuxWebOnly: string[];
  desktopTauriOnly: string[];
  forbidden: {
    desktopEmbeddedTauri: string[];
    desktopLanWebui: string[];
  };
};

export type NexusHubContractRegistry = {
  schemaVersion: number;
  hostSurfaces: HostSurface[];
  capabilities: string[];
  capabilitiesByHostSurface: Record<HostSurface, string[]>;
  visual: ContractVisual;
  actions: ContractAction[];
  dtoCatalog: Record<string, ContractDtoCatalogEntry>;
};

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isStringArray(value: unknown): value is string[] {
  return Array.isArray(value) && value.every((item) => typeof item === "string" && item.trim().length > 0);
}

function assertContractRegistry(value: unknown): asserts value is NexusHubContractRegistry {
  if (!isObject(value)) {
    throw new Error("contract registry must be an object");
  }
  for (const key of ["schemaVersion", "hostSurfaces", "capabilities", "capabilitiesByHostSurface", "visual", "actions", "dtoCatalog"]) {
    if (!(key in value)) {
      throw new Error(`contract registry missing ${key}`);
    }
  }
  if (value.schemaVersion !== 1) {
    throw new Error("contract registry schemaVersion must be 1");
  }
  if (!isStringArray(value.hostSurfaces)) {
    throw new Error("contract registry hostSurfaces must be a non-empty string array");
  }
  if (!isStringArray(value.capabilities)) {
    throw new Error("contract registry capabilities must be a non-empty string array");
  }
  if (!isObject(value.capabilitiesByHostSurface)) {
    throw new Error("contract registry capabilitiesByHostSurface must be an object");
  }
  for (const surface of value.hostSurfaces) {
    if (!isStringArray(value.capabilitiesByHostSurface[surface])) {
      throw new Error(`contract registry missing capabilities for ${surface}`);
    }
  }
  if (!isObject(value.visual) || !isObject(value.visual.forbidden)) {
    throw new Error("contract registry visual rules must be objects");
  }
  for (const key of ["navigation", "corePanelTitles", "linuxWebOnly", "desktopTauriOnly"]) {
    if (!isStringArray(value.visual[key])) {
      throw new Error(`contract registry visual.${key} must be a string array`);
    }
  }
  if (!isObject(value.visual.actionLabels) || !isObject(value.visual.disabledStates)) {
    throw new Error("contract registry visual labels and disabled states must be objects");
  }
  if (!isStringArray(value.visual.forbidden.desktopEmbeddedTauri) || !isStringArray(value.visual.forbidden.desktopLanWebui)) {
    throw new Error("contract registry forbidden visual rules must cover desktop host surfaces");
  }
  if (!Array.isArray(value.actions)) {
    throw new Error("contract registry actions must be an array");
  }
  if (!isObject(value.dtoCatalog)) {
    throw new Error("contract registry dtoCatalog must be an object");
  }
  for (const [name, entry] of Object.entries(value.dtoCatalog)) {
    if (!isObject(entry)) {
      throw new Error(`contract registry dtoCatalog.${name} must be an object`);
    }
    for (const key of ["core", "webui"]) {
      if (typeof entry[key] !== "string" || entry[key].trim().length === 0) {
        throw new Error(`contract registry dtoCatalog.${name}.${key} must be a non-empty string`);
      }
    }
  }
  for (const action of value.actions) {
    if (!isObject(action)) {
      throw new Error("contract registry action must be an object");
    }
    const id = action.id;
    if (typeof id !== "string" || id.trim().length === 0) {
      throw new Error("contract registry action id must be a non-empty string");
    }
    if (action.kind !== "rpc" && action.kind !== "transport") {
      throw new Error(`contract registry action ${id} has unsupported kind`);
    }
    if (action.scope !== "shared" && action.scope !== "host_only" && action.scope !== "transport") {
      throw new Error(`contract registry action ${id} has unsupported scope`);
    }
    if (typeof action.coreUseCase !== "string" || action.coreUseCase.trim().length === 0) {
      throw new Error(`contract registry action ${id} must declare coreUseCase`);
    }
    if (action.scope === "shared") {
      for (const key of ["linuxRpc", "tauriCommand", "webuiWrapper", "dtoOwner", "requestDto", "responseDto"]) {
        if (typeof action[key] !== "string" || action[key].trim().length === 0) {
          throw new Error(`shared contract action ${id} must declare ${key}`);
        }
      }
    }
    if (action.scope === "transport") {
      for (const key of ["webuiWrapper", "dtoOwner", "requestDto", "responseDto"]) {
        if (typeof action[key] !== "string" || action[key].trim().length === 0) {
          throw new Error(`transport contract action ${id} must declare ${key}`);
        }
      }
    }
    for (const key of ["requestDto", "responseDto"]) {
      const dtoName = action[key];
      if (typeof dtoName === "string" && !(dtoName in value.dtoCatalog)) {
        throw new Error(`contract registry action ${id} references unknown ${key} ${dtoName}`);
      }
    }
    if (action.scope === "host_only" && (typeof action.hostOnlyReason !== "string" || action.hostOnlyReason.trim().length === 0)) {
      throw new Error(`host-only contract action ${id} must declare hostOnlyReason`);
    }
  }
}

assertContractRegistry(registry);

export const contractRegistry = registry;

export const contractHostSurfaces = contractRegistry.hostSurfaces;
export const contractCapabilities = contractRegistry.capabilities;
export const contractCapabilitiesByHostSurface = contractRegistry.capabilitiesByHostSurface;
export const contractVisual = contractRegistry.visual;
export const contractActions = contractRegistry.actions;
export const contractDtoCatalog = contractRegistry.dtoCatalog;
export const sharedContractActions = contractActions.filter((action) => action.scope === "shared");
export const hostOnlyContractActions = contractActions.filter((action) => action.scope === "host_only");
